### DISCUSSION TITLE
Eager-load related elements (across all sites) on auto-injected entry`s template

### DISCUSSION THREAD

**Discussion by richhayler** [craftcms/cms#11393] [Q&A] [answered]
Is there a way to eager-load related elements (entries field), across all sites for an auto-injected entry's template?

Running the following code in an entry template to eager load the related entries just returns the items related to the current site. Is there a way I can eager-load the entry's relates entries across all sites or do I have to  set this up seperately as shown below the first example? 

```
{% do craft.app.elements.eagerLoadElements(
	className(entry),
	[entry],
	['myEntriesField', 'someAssetField']
) %}

{# this only shows the entries relate to the entry's site #}
{% for item in entry.myEntries.all() %}
    {{ item.title }}
{% endear %}
```

Is the solution to remove `myEntriesField` from the initial query and then create a seperate collection e.g.

```
{% set myEntries = entry.blocks
	.site('*')
	.collect() %}
```
And then iterate over the array

```
{# this will show all entries (across all sites) for the entry #}
{% for item in myEntries %}
    {{ item.title }}
{% endear %}
```

https://github.com/craftcms/cms/blob/main/src/services/Elements.php#L2219-L2230

**Comment by brandonkelly** [comment, marked as answer]
Relational fields will query for related elements in the same site as the source element by default (or in the Target Site, if that setting is set), whether you’re eager-loading or not, and in either case you can change that behavior by overriding the `site`/`siteId` param.

Without eager-loading:

```twig
{% set relatedEntries = entry.myEntriesField
  .site('*')
  .collect() %}
```

With eager-loading:
```twig
{% set sourceEntries = craft.entries()
  .section('foo')
  .with([
    ['myEntriesField', {site: '*'}],
  ])
  .collect() %}

{% for sourceEntry in sourceEntries %}
  {% set relatedEntries = sourceEntry.myEntriesField.collect() %}
  ...
{% endfor %}
```

Note that there is almost never a good reason to call `craft.app.elements.eagerLoadElements()` from Twig, and certainly never a reason to call it for only a single source element (`[entry]`). The point of eager-loading is to avoid `n+1` queries, when you are querying for multiple elements, and each of those elements will have their own nested elements that also need to be loaded.

For example, if you need to fetch a related asset off of each of your related elements, you could eager-load them when first fetching the related entries:

```twig
{% set relatedEntries = entry.myEntriesField
  .site('*')
  .with(['myAssetsField'])
  .collect() %}
```

**Reply by richhayler** [reply]
Thank you @brandonkelly for your thoroughly informative and  comprehensive explanation. I really appreciate the examples you provided which have helped me immensely! Thank you.

**Reply by brandonkelly** [reply]
Glad it helped! I should have mentioned, you may want to couple `site('*')` with [`unique()`](https://craftcms.com/docs/4.x/entries.html#unique) to avoid getting the same related elements back multiple times, if any exist across multiple sites. You can also add [`preferSites([currentSite.handle])`](https://craftcms.com/docs/4.x/entries.html#prefersites) to ensure elements that do exist for the current site will be returned in that site.
